{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "source('/hpc/users/torred23/pipelines/support/init.R')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "aggregate_counts <- function(infiles, method, txOut=TRUE, genome='hg38', name_func=function(x) { x_split <- strsplit(x, '/')[[1]]; x_name <- x_split[length(x_split)-1]; return(x_name); }, star_col=2) {\n",
    "\n",
    "    # Add names\n",
    "    names(infiles) <- sapply(infiles, name_func)\n",
    "\n",
    "    # STAR\n",
    "    if (method == 'star') {\n",
    "         \n",
    "        # Read and concatenate\n",
    "        concatenated_dataframe <- do.call('rbind', sapply(names(infiles), function(x) {df <- as.data.frame(data.table::fread(infiles[x])); df <- df[grepl('ENSG*', df$V1),]; colnames(df)[1] <- 'ensembl_gene_id'; colnames(df)[star_col] <- 'counts'; df$sample <- x; df}, simplify=FALSE))\n",
    "        \n",
    "        # Cast\n",
    "        result <- reshape2::dcast(ensembl_gene_id ~ sample, data = concatenated_dataframe, value.var='counts', fun.aggregate=sum)\n",
    "               \n",
    "    # Kallisto and Salmon\n",
    "    } else if (method %in% c('kallisto', 'salmon')) {\n",
    "      \n",
    "        # Get tx2g\n",
    "        tx2g <- data.table::fread(Sys.glob(glue::glue('/sc/hydra/projects/GuccioneLab/genome-indices/{genome}/ensembl/*-biomart.txt')))\n",
    "\n",
    "        # Tximport\n",
    "        result <- tximport::tximport(infiles, type=method, txOut=txOut, tx2g=tx2g[,c('Transcript stable ID version', 'Gene stable ID')])\n",
    "\n",
    "    }\n",
    "    \n",
    "    # Return\n",
    "    return(result)\n",
    "        \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 25</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ensembl_gene_id</th><th scope=col>1A_S1_R1_001</th><th scope=col>1B_S2_R1_001</th><th scope=col>1C_S3_R1_001</th><th scope=col>1E_S4_R1_001</th><th scope=col>1F_S5_R1_001</th><th scope=col>1G_S6_R1_001</th><th scope=col>3A_S7_R1_001</th><th scope=col>3B_S8_R1_001</th><th scope=col>3C_S9_R1_001</th><th scope=col>⋯</th><th scope=col>5C_S15_R1_001</th><th scope=col>5D_S16_R1_001</th><th scope=col>5E_S17_R1_001</th><th scope=col>5G_S18_R1_001</th><th scope=col>7A_S19_R1_001</th><th scope=col>7B_S20_R1_001</th><th scope=col>7C_S21_R1_001</th><th scope=col>7D_S22_R1_001</th><th scope=col>7E_S23_R1_001</th><th scope=col>7F_S24_R1_001</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000000003</td><td>   42</td><td>   42</td><td>   26</td><td>   22</td><td>   38</td><td>   16</td><td>   30</td><td>   18</td><td>   28</td><td>⋯</td><td>   16</td><td>   14</td><td>   10</td><td>   24</td><td>   16</td><td>   12</td><td>   24</td><td>   18</td><td>   38</td><td>   24</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000000005</td><td>    8</td><td>    0</td><td>    2</td><td>    0</td><td>    0</td><td>    0</td><td>    0</td><td>    0</td><td>    0</td><td>⋯</td><td>    2</td><td>    0</td><td>    0</td><td>    0</td><td>    0</td><td>    0</td><td>    0</td><td>    0</td><td>    0</td><td>    0</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENSG00000000419</td><td> 1622</td><td> 1542</td><td> 1042</td><td> 1530</td><td> 1494</td><td> 1132</td><td> 1686</td><td> 1808</td><td> 1706</td><td>⋯</td><td> 1456</td><td> 1270</td><td> 1120</td><td> 1302</td><td> 1376</td><td> 1208</td><td> 1472</td><td> 1710</td><td> 1574</td><td> 1596</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENSG00000000457</td><td>  816</td><td>  904</td><td>  596</td><td>  886</td><td>  780</td><td>  758</td><td>  634</td><td>  422</td><td>  802</td><td>⋯</td><td>  710</td><td>  594</td><td>  524</td><td>  678</td><td>  598</td><td>  560</td><td>  612</td><td>  750</td><td>  668</td><td>  848</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENSG00000000460</td><td>  164</td><td>  190</td><td>  140</td><td>  162</td><td>  140</td><td>  102</td><td>  110</td><td>   98</td><td>  162</td><td>⋯</td><td>  144</td><td>  114</td><td>  112</td><td>  150</td><td>  262</td><td>  132</td><td>  178</td><td>  232</td><td>  190</td><td>  248</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENSG00000000938</td><td>21454</td><td>29264</td><td>15542</td><td>31384</td><td>24234</td><td>25276</td><td>26542</td><td>24120</td><td>30468</td><td>⋯</td><td>22944</td><td>20328</td><td>20232</td><td>22030</td><td>17380</td><td>18940</td><td>21722</td><td>26046</td><td>29584</td><td>26898</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 25\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & ensembl\\_gene\\_id & 1A\\_S1\\_R1\\_001 & 1B\\_S2\\_R1\\_001 & 1C\\_S3\\_R1\\_001 & 1E\\_S4\\_R1\\_001 & 1F\\_S5\\_R1\\_001 & 1G\\_S6\\_R1\\_001 & 3A\\_S7\\_R1\\_001 & 3B\\_S8\\_R1\\_001 & 3C\\_S9\\_R1\\_001 & ⋯ & 5C\\_S15\\_R1\\_001 & 5D\\_S16\\_R1\\_001 & 5E\\_S17\\_R1\\_001 & 5G\\_S18\\_R1\\_001 & 7A\\_S19\\_R1\\_001 & 7B\\_S20\\_R1\\_001 & 7C\\_S21\\_R1\\_001 & 7D\\_S22\\_R1\\_001 & 7E\\_S23\\_R1\\_001 & 7F\\_S24\\_R1\\_001\\\\\n",
       "  & <chr> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & ⋯ & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000000003 &    42 &    42 &    26 &    22 &    38 &    16 &    30 &    18 &    28 & ⋯ &    16 &    14 &    10 &    24 &    16 &    12 &    24 &    18 &    38 &    24\\\\\n",
       "\t2 & ENSG00000000005 &     8 &     0 &     2 &     0 &     0 &     0 &     0 &     0 &     0 & ⋯ &     2 &     0 &     0 &     0 &     0 &     0 &     0 &     0 &     0 &     0\\\\\n",
       "\t3 & ENSG00000000419 &  1622 &  1542 &  1042 &  1530 &  1494 &  1132 &  1686 &  1808 &  1706 & ⋯ &  1456 &  1270 &  1120 &  1302 &  1376 &  1208 &  1472 &  1710 &  1574 &  1596\\\\\n",
       "\t4 & ENSG00000000457 &   816 &   904 &   596 &   886 &   780 &   758 &   634 &   422 &   802 & ⋯ &   710 &   594 &   524 &   678 &   598 &   560 &   612 &   750 &   668 &   848\\\\\n",
       "\t5 & ENSG00000000460 &   164 &   190 &   140 &   162 &   140 &   102 &   110 &    98 &   162 & ⋯ &   144 &   114 &   112 &   150 &   262 &   132 &   178 &   232 &   190 &   248\\\\\n",
       "\t6 & ENSG00000000938 & 21454 & 29264 & 15542 & 31384 & 24234 & 25276 & 26542 & 24120 & 30468 & ⋯ & 22944 & 20328 & 20232 & 22030 & 17380 & 18940 & 21722 & 26046 & 29584 & 26898\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 25\n",
       "\n",
       "| <!--/--> | ensembl_gene_id &lt;chr&gt; | 1A_S1_R1_001 &lt;int&gt; | 1B_S2_R1_001 &lt;int&gt; | 1C_S3_R1_001 &lt;int&gt; | 1E_S4_R1_001 &lt;int&gt; | 1F_S5_R1_001 &lt;int&gt; | 1G_S6_R1_001 &lt;int&gt; | 3A_S7_R1_001 &lt;int&gt; | 3B_S8_R1_001 &lt;int&gt; | 3C_S9_R1_001 &lt;int&gt; | ⋯ ⋯ | 5C_S15_R1_001 &lt;int&gt; | 5D_S16_R1_001 &lt;int&gt; | 5E_S17_R1_001 &lt;int&gt; | 5G_S18_R1_001 &lt;int&gt; | 7A_S19_R1_001 &lt;int&gt; | 7B_S20_R1_001 &lt;int&gt; | 7C_S21_R1_001 &lt;int&gt; | 7D_S22_R1_001 &lt;int&gt; | 7E_S23_R1_001 &lt;int&gt; | 7F_S24_R1_001 &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENSG00000000003 |    42 |    42 |    26 |    22 |    38 |    16 |    30 |    18 |    28 | ⋯ |    16 |    14 |    10 |    24 |    16 |    12 |    24 |    18 |    38 |    24 |\n",
       "| 2 | ENSG00000000005 |     8 |     0 |     2 |     0 |     0 |     0 |     0 |     0 |     0 | ⋯ |     2 |     0 |     0 |     0 |     0 |     0 |     0 |     0 |     0 |     0 |\n",
       "| 3 | ENSG00000000419 |  1622 |  1542 |  1042 |  1530 |  1494 |  1132 |  1686 |  1808 |  1706 | ⋯ |  1456 |  1270 |  1120 |  1302 |  1376 |  1208 |  1472 |  1710 |  1574 |  1596 |\n",
       "| 4 | ENSG00000000457 |   816 |   904 |   596 |   886 |   780 |   758 |   634 |   422 |   802 | ⋯ |   710 |   594 |   524 |   678 |   598 |   560 |   612 |   750 |   668 |   848 |\n",
       "| 5 | ENSG00000000460 |   164 |   190 |   140 |   162 |   140 |   102 |   110 |    98 |   162 | ⋯ |   144 |   114 |   112 |   150 |   262 |   132 |   178 |   232 |   190 |   248 |\n",
       "| 6 | ENSG00000000938 | 21454 | 29264 | 15542 | 31384 | 24234 | 25276 | 26542 | 24120 | 30468 | ⋯ | 22944 | 20328 | 20232 | 22030 | 17380 | 18940 | 21722 | 26046 | 29584 | 26898 |\n",
       "\n"
      ],
      "text/plain": [
       "  ensembl_gene_id 1A_S1_R1_001 1B_S2_R1_001 1C_S3_R1_001 1E_S4_R1_001\n",
       "1 ENSG00000000003    42           42           26           22       \n",
       "2 ENSG00000000005     8            0            2            0       \n",
       "3 ENSG00000000419  1622         1542         1042         1530       \n",
       "4 ENSG00000000457   816          904          596          886       \n",
       "5 ENSG00000000460   164          190          140          162       \n",
       "6 ENSG00000000938 21454        29264        15542        31384       \n",
       "  1F_S5_R1_001 1G_S6_R1_001 3A_S7_R1_001 3B_S8_R1_001 3C_S9_R1_001 ⋯\n",
       "1    38           16           30           18           28        ⋯\n",
       "2     0            0            0            0            0        ⋯\n",
       "3  1494         1132         1686         1808         1706        ⋯\n",
       "4   780          758          634          422          802        ⋯\n",
       "5   140          102          110           98          162        ⋯\n",
       "6 24234        25276        26542        24120        30468        ⋯\n",
       "  5C_S15_R1_001 5D_S16_R1_001 5E_S17_R1_001 5G_S18_R1_001 7A_S19_R1_001\n",
       "1    16            14            10            24            16        \n",
       "2     2             0             0             0             0        \n",
       "3  1456          1270          1120          1302          1376        \n",
       "4   710           594           524           678           598        \n",
       "5   144           114           112           150           262        \n",
       "6 22944         20328         20232         22030         17380        \n",
       "  7B_S20_R1_001 7C_S21_R1_001 7D_S22_R1_001 7E_S23_R1_001 7F_S24_R1_001\n",
       "1    12            24            18            38            24        \n",
       "2     0             0             0             0             0        \n",
       "3  1208          1472          1710          1574          1596        \n",
       "4   560           612           750           668           848        \n",
       "5   132           178           232           190           248        \n",
       "6 18940         21722         26046         29584         26898        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "infiles <- Sys.glob('/hpc/users/torred23/pipelines/GuccioneLab/covid-tcell/hydra/s2-alignment.dir/star/*/*out.tab')\n",
    "star_dataframe <- aggregate_counts(infiles, method='star', genome='hg38')\n",
    "head(star_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "10 \n",
      "11 \n",
      "12 \n",
      "13 \n",
      "14 \n",
      "15 \n",
      "16 \n",
      "17 \n",
      "18 \n",
      "19 \n",
      "20 \n",
      "21 \n",
      "22 \n",
      "23 \n",
      "24 \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "infiles <- Sys.glob('/hpc/users/torred23/pipelines/GuccioneLab/covid-tcell/hydra/s2-alignment.dir/kallisto/*/abundance.h5')\n",
    "kallisto_txi <- aggregate_counts(infiles, method='kallisto', txOut=FALSE, genome='hg38')\n",
    "head(kallisto_txi$counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "10 \n",
      "11 \n",
      "12 \n",
      "13 \n",
      "14 \n",
      "15 \n",
      "16 \n",
      "17 \n",
      "18 \n",
      "19 \n",
      "20 \n",
      "21 \n",
      "22 \n",
      "23 \n",
      "24 \n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 6 × 24 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>1A_S1_R1_001</th><th scope=col>1B_S2_R1_001</th><th scope=col>1C_S3_R1_001</th><th scope=col>1E_S4_R1_001</th><th scope=col>1F_S5_R1_001</th><th scope=col>1G_S6_R1_001</th><th scope=col>3A_S7_R1_001</th><th scope=col>3B_S8_R1_001</th><th scope=col>3C_S9_R1_001</th><th scope=col>3D_S10_R1_001</th><th scope=col>⋯</th><th scope=col>5C_S15_R1_001</th><th scope=col>5D_S16_R1_001</th><th scope=col>5E_S17_R1_001</th><th scope=col>5G_S18_R1_001</th><th scope=col>7A_S19_R1_001</th><th scope=col>7B_S20_R1_001</th><th scope=col>7C_S21_R1_001</th><th scope=col>7D_S22_R1_001</th><th scope=col>7E_S23_R1_001</th><th scope=col>7F_S24_R1_001</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENST00000434970.2</th><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENST00000415118.1</th><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENST00000448914.1</th><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENST00000631435.1</th><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENST00000390583.1</th><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENST00000431440.2</th><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 24 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & 1A\\_S1\\_R1\\_001 & 1B\\_S2\\_R1\\_001 & 1C\\_S3\\_R1\\_001 & 1E\\_S4\\_R1\\_001 & 1F\\_S5\\_R1\\_001 & 1G\\_S6\\_R1\\_001 & 3A\\_S7\\_R1\\_001 & 3B\\_S8\\_R1\\_001 & 3C\\_S9\\_R1\\_001 & 3D\\_S10\\_R1\\_001 & ⋯ & 5C\\_S15\\_R1\\_001 & 5D\\_S16\\_R1\\_001 & 5E\\_S17\\_R1\\_001 & 5G\\_S18\\_R1\\_001 & 7A\\_S19\\_R1\\_001 & 7B\\_S20\\_R1\\_001 & 7C\\_S21\\_R1\\_001 & 7D\\_S22\\_R1\\_001 & 7E\\_S23\\_R1\\_001 & 7F\\_S24\\_R1\\_001\\\\\n",
       "\\hline\n",
       "\tENST00000434970.2 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tENST00000415118.1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tENST00000448914.1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tENST00000631435.1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tENST00000390583.1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tENST00000431440.2 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 24 of type dbl\n",
       "\n",
       "| <!--/--> | 1A_S1_R1_001 | 1B_S2_R1_001 | 1C_S3_R1_001 | 1E_S4_R1_001 | 1F_S5_R1_001 | 1G_S6_R1_001 | 3A_S7_R1_001 | 3B_S8_R1_001 | 3C_S9_R1_001 | 3D_S10_R1_001 | ⋯ | 5C_S15_R1_001 | 5D_S16_R1_001 | 5E_S17_R1_001 | 5G_S18_R1_001 | 7A_S19_R1_001 | 7B_S20_R1_001 | 7C_S21_R1_001 | 7D_S22_R1_001 | 7E_S23_R1_001 | 7F_S24_R1_001 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ENST00000434970.2 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| ENST00000415118.1 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| ENST00000448914.1 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| ENST00000631435.1 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| ENST00000390583.1 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| ENST00000431440.2 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "                  1A_S1_R1_001 1B_S2_R1_001 1C_S3_R1_001 1E_S4_R1_001\n",
       "ENST00000434970.2 0            0            0            0           \n",
       "ENST00000415118.1 0            0            0            0           \n",
       "ENST00000448914.1 0            0            0            0           \n",
       "ENST00000631435.1 0            0            0            0           \n",
       "ENST00000390583.1 0            0            0            0           \n",
       "ENST00000431440.2 0            0            0            0           \n",
       "                  1F_S5_R1_001 1G_S6_R1_001 3A_S7_R1_001 3B_S8_R1_001\n",
       "ENST00000434970.2 0            0            0            0           \n",
       "ENST00000415118.1 0            0            0            0           \n",
       "ENST00000448914.1 0            0            0            0           \n",
       "ENST00000631435.1 0            0            0            0           \n",
       "ENST00000390583.1 0            0            0            0           \n",
       "ENST00000431440.2 0            0            0            0           \n",
       "                  3C_S9_R1_001 3D_S10_R1_001 ⋯ 5C_S15_R1_001 5D_S16_R1_001\n",
       "ENST00000434970.2 0            0             ⋯ 0             0            \n",
       "ENST00000415118.1 0            0             ⋯ 0             0            \n",
       "ENST00000448914.1 0            0             ⋯ 0             0            \n",
       "ENST00000631435.1 0            0             ⋯ 0             0            \n",
       "ENST00000390583.1 0            0             ⋯ 0             0            \n",
       "ENST00000431440.2 0            0             ⋯ 0             0            \n",
       "                  5E_S17_R1_001 5G_S18_R1_001 7A_S19_R1_001 7B_S20_R1_001\n",
       "ENST00000434970.2 0             0             0             0            \n",
       "ENST00000415118.1 0             0             0             0            \n",
       "ENST00000448914.1 0             0             0             0            \n",
       "ENST00000631435.1 0             0             0             0            \n",
       "ENST00000390583.1 0             0             0             0            \n",
       "ENST00000431440.2 0             0             0             0            \n",
       "                  7C_S21_R1_001 7D_S22_R1_001 7E_S23_R1_001 7F_S24_R1_001\n",
       "ENST00000434970.2 0             0             0             0            \n",
       "ENST00000415118.1 0             0             0             0            \n",
       "ENST00000448914.1 0             0             0             0            \n",
       "ENST00000631435.1 0             0             0             0            \n",
       "ENST00000390583.1 0             0             0             0            \n",
       "ENST00000431440.2 0             0             0             0            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "infiles <- Sys.glob('/hpc/users/torred23/pipelines/GuccioneLab/covid-tcell/hydra/s2-alignment.dir/salmon/*/quant.sf')\n",
    "salmon_txi <- aggregate_counts(infiles, method='salmon', txOut=Fgenome='hg38')\n",
    "head(salmon_txi$counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get files\n",
    "infiles <- Sys.glob(file.path(results_dir, '*', '*abundance.h5'))\n",
    "\n",
    "# Add names\n",
    "names(infiles) <- sapply(infiles, name_func)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import\n",
    "require(tximport)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome='hg38'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get tx2g\n",
    "tx2g <- data.table::fread(Sys.glob(glue::glue('/sc/hydra/projects/GuccioneLab/genome-indices/{genome}/ensembl/*-biomart.txt')))[,c(1, 4)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 6 × 24 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>1A_S1_R1_001</th><th scope=col>1B_S2_R1_001</th><th scope=col>1C_S3_R1_001</th><th scope=col>1E_S4_R1_001</th><th scope=col>1F_S5_R1_001</th><th scope=col>1G_S6_R1_001</th><th scope=col>3A_S7_R1_001</th><th scope=col>3B_S8_R1_001</th><th scope=col>3C_S9_R1_001</th><th scope=col>3D_S10_R1_001</th><th scope=col>⋯</th><th scope=col>5C_S15_R1_001</th><th scope=col>5D_S16_R1_001</th><th scope=col>5E_S17_R1_001</th><th scope=col>5G_S18_R1_001</th><th scope=col>7A_S19_R1_001</th><th scope=col>7B_S20_R1_001</th><th scope=col>7C_S21_R1_001</th><th scope=col>7D_S22_R1_001</th><th scope=col>7E_S23_R1_001</th><th scope=col>7F_S24_R1_001</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000000003</th><td>   51.05216</td><td>   40.79417</td><td>  32.23023</td><td>   33.15891</td><td>   44.46954</td><td>   28.42961</td><td>   23.3118</td><td>   32.25710</td><td>   27.76952</td><td>   33.67071</td><td>⋯</td><td>   25.53284</td><td>  16.51529</td><td>  12.08209</td><td>   23.44132</td><td>  18.27322</td><td>  14.45599</td><td>   17.61952</td><td>   22.95925</td><td>   37.04031</td><td>   35.04609</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000005</th><td>    2.00000</td><td>    0.00000</td><td>   1.00000</td><td>    0.00000</td><td>    0.00000</td><td>    0.00000</td><td>    0.0000</td><td>    0.00000</td><td>    0.00000</td><td>    1.00000</td><td>⋯</td><td>    1.00000</td><td>   0.00000</td><td>   0.00000</td><td>    0.00000</td><td>   0.00000</td><td>   0.00000</td><td>    0.00000</td><td>    0.00000</td><td>    0.00000</td><td>    0.00000</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000419</th><td>  750.00000</td><td>  727.00000</td><td> 496.00000</td><td>  721.00000</td><td>  705.00000</td><td>  542.00000</td><td>  787.0000</td><td>  853.00000</td><td>  814.00000</td><td>  697.00000</td><td>⋯</td><td>  683.00000</td><td> 566.00000</td><td> 523.00000</td><td>  627.00000</td><td> 646.00000</td><td> 561.00000</td><td>  702.00000</td><td>  826.00000</td><td>  730.00000</td><td>  755.00000</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000457</th><td>  603.07870</td><td>  646.72254</td><td> 448.09453</td><td>  682.70977</td><td>  558.08977</td><td>  519.49518</td><td>  521.2949</td><td>  431.25536</td><td>  593.09671</td><td>  780.54047</td><td>⋯</td><td>  491.88729</td><td> 469.26460</td><td> 381.83138</td><td>  522.13590</td><td> 474.65664</td><td> 436.29466</td><td>  457.43441</td><td>  590.91226</td><td>  538.29130</td><td>  637.10988</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000460</th><td>  145.13014</td><td>  135.47068</td><td>  87.20398</td><td>  119.15808</td><td>  117.41435</td><td>   92.39445</td><td>   89.9817</td><td>   98.06563</td><td>  107.90646</td><td>  134.14691</td><td>⋯</td><td>   98.03182</td><td>  79.98037</td><td>  75.98264</td><td>   95.93890</td><td> 151.43542</td><td>  84.70761</td><td>  132.63278</td><td>  164.68933</td><td>  123.11935</td><td>  161.07990</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000938</th><td>10498.00000</td><td>14261.00000</td><td>7633.00000</td><td>15374.00000</td><td>11941.00000</td><td>12390.00000</td><td>12951.0000</td><td>11913.00000</td><td>14857.00000</td><td>10971.00000</td><td>⋯</td><td>11255.00000</td><td>9963.00000</td><td>9904.00000</td><td>10703.00000</td><td>8539.00000</td><td>9219.00000</td><td>10615.00000</td><td>12813.00000</td><td>14505.00000</td><td>13227.00000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 24 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & 1A\\_S1\\_R1\\_001 & 1B\\_S2\\_R1\\_001 & 1C\\_S3\\_R1\\_001 & 1E\\_S4\\_R1\\_001 & 1F\\_S5\\_R1\\_001 & 1G\\_S6\\_R1\\_001 & 3A\\_S7\\_R1\\_001 & 3B\\_S8\\_R1\\_001 & 3C\\_S9\\_R1\\_001 & 3D\\_S10\\_R1\\_001 & ⋯ & 5C\\_S15\\_R1\\_001 & 5D\\_S16\\_R1\\_001 & 5E\\_S17\\_R1\\_001 & 5G\\_S18\\_R1\\_001 & 7A\\_S19\\_R1\\_001 & 7B\\_S20\\_R1\\_001 & 7C\\_S21\\_R1\\_001 & 7D\\_S22\\_R1\\_001 & 7E\\_S23\\_R1\\_001 & 7F\\_S24\\_R1\\_001\\\\\n",
       "\\hline\n",
       "\tENSG00000000003 &    51.05216 &    40.79417 &   32.23023 &    33.15891 &    44.46954 &    28.42961 &    23.3118 &    32.25710 &    27.76952 &    33.67071 & ⋯ &    25.53284 &   16.51529 &   12.08209 &    23.44132 &   18.27322 &   14.45599 &    17.61952 &    22.95925 &    37.04031 &    35.04609\\\\\n",
       "\tENSG00000000005 &     2.00000 &     0.00000 &    1.00000 &     0.00000 &     0.00000 &     0.00000 &     0.0000 &     0.00000 &     0.00000 &     1.00000 & ⋯ &     1.00000 &    0.00000 &    0.00000 &     0.00000 &    0.00000 &    0.00000 &     0.00000 &     0.00000 &     0.00000 &     0.00000\\\\\n",
       "\tENSG00000000419 &   750.00000 &   727.00000 &  496.00000 &   721.00000 &   705.00000 &   542.00000 &   787.0000 &   853.00000 &   814.00000 &   697.00000 & ⋯ &   683.00000 &  566.00000 &  523.00000 &   627.00000 &  646.00000 &  561.00000 &   702.00000 &   826.00000 &   730.00000 &   755.00000\\\\\n",
       "\tENSG00000000457 &   603.07870 &   646.72254 &  448.09453 &   682.70977 &   558.08977 &   519.49518 &   521.2949 &   431.25536 &   593.09671 &   780.54047 & ⋯ &   491.88729 &  469.26460 &  381.83138 &   522.13590 &  474.65664 &  436.29466 &   457.43441 &   590.91226 &   538.29130 &   637.10988\\\\\n",
       "\tENSG00000000460 &   145.13014 &   135.47068 &   87.20398 &   119.15808 &   117.41435 &    92.39445 &    89.9817 &    98.06563 &   107.90646 &   134.14691 & ⋯ &    98.03182 &   79.98037 &   75.98264 &    95.93890 &  151.43542 &   84.70761 &   132.63278 &   164.68933 &   123.11935 &   161.07990\\\\\n",
       "\tENSG00000000938 & 10498.00000 & 14261.00000 & 7633.00000 & 15374.00000 & 11941.00000 & 12390.00000 & 12951.0000 & 11913.00000 & 14857.00000 & 10971.00000 & ⋯ & 11255.00000 & 9963.00000 & 9904.00000 & 10703.00000 & 8539.00000 & 9219.00000 & 10615.00000 & 12813.00000 & 14505.00000 & 13227.00000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 24 of type dbl\n",
       "\n",
       "| <!--/--> | 1A_S1_R1_001 | 1B_S2_R1_001 | 1C_S3_R1_001 | 1E_S4_R1_001 | 1F_S5_R1_001 | 1G_S6_R1_001 | 3A_S7_R1_001 | 3B_S8_R1_001 | 3C_S9_R1_001 | 3D_S10_R1_001 | ⋯ | 5C_S15_R1_001 | 5D_S16_R1_001 | 5E_S17_R1_001 | 5G_S18_R1_001 | 7A_S19_R1_001 | 7B_S20_R1_001 | 7C_S21_R1_001 | 7D_S22_R1_001 | 7E_S23_R1_001 | 7F_S24_R1_001 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ENSG00000000003 |    51.05216 |    40.79417 |   32.23023 |    33.15891 |    44.46954 |    28.42961 |    23.3118 |    32.25710 |    27.76952 |    33.67071 | ⋯ |    25.53284 |   16.51529 |   12.08209 |    23.44132 |   18.27322 |   14.45599 |    17.61952 |    22.95925 |    37.04031 |    35.04609 |\n",
       "| ENSG00000000005 |     2.00000 |     0.00000 |    1.00000 |     0.00000 |     0.00000 |     0.00000 |     0.0000 |     0.00000 |     0.00000 |     1.00000 | ⋯ |     1.00000 |    0.00000 |    0.00000 |     0.00000 |    0.00000 |    0.00000 |     0.00000 |     0.00000 |     0.00000 |     0.00000 |\n",
       "| ENSG00000000419 |   750.00000 |   727.00000 |  496.00000 |   721.00000 |   705.00000 |   542.00000 |   787.0000 |   853.00000 |   814.00000 |   697.00000 | ⋯ |   683.00000 |  566.00000 |  523.00000 |   627.00000 |  646.00000 |  561.00000 |   702.00000 |   826.00000 |   730.00000 |   755.00000 |\n",
       "| ENSG00000000457 |   603.07870 |   646.72254 |  448.09453 |   682.70977 |   558.08977 |   519.49518 |   521.2949 |   431.25536 |   593.09671 |   780.54047 | ⋯ |   491.88729 |  469.26460 |  381.83138 |   522.13590 |  474.65664 |  436.29466 |   457.43441 |   590.91226 |   538.29130 |   637.10988 |\n",
       "| ENSG00000000460 |   145.13014 |   135.47068 |   87.20398 |   119.15808 |   117.41435 |    92.39445 |    89.9817 |    98.06563 |   107.90646 |   134.14691 | ⋯ |    98.03182 |   79.98037 |   75.98264 |    95.93890 |  151.43542 |   84.70761 |   132.63278 |   164.68933 |   123.11935 |   161.07990 |\n",
       "| ENSG00000000938 | 10498.00000 | 14261.00000 | 7633.00000 | 15374.00000 | 11941.00000 | 12390.00000 | 12951.0000 | 11913.00000 | 14857.00000 | 10971.00000 | ⋯ | 11255.00000 | 9963.00000 | 9904.00000 | 10703.00000 | 8539.00000 | 9219.00000 | 10615.00000 | 12813.00000 | 14505.00000 | 13227.00000 |\n",
       "\n"
      ],
      "text/plain": [
       "                1A_S1_R1_001 1B_S2_R1_001 1C_S3_R1_001 1E_S4_R1_001\n",
       "ENSG00000000003    51.05216     40.79417    32.23023      33.15891 \n",
       "ENSG00000000005     2.00000      0.00000     1.00000       0.00000 \n",
       "ENSG00000000419   750.00000    727.00000   496.00000     721.00000 \n",
       "ENSG00000000457   603.07870    646.72254   448.09453     682.70977 \n",
       "ENSG00000000460   145.13014    135.47068    87.20398     119.15808 \n",
       "ENSG00000000938 10498.00000  14261.00000  7633.00000   15374.00000 \n",
       "                1F_S5_R1_001 1G_S6_R1_001 3A_S7_R1_001 3B_S8_R1_001\n",
       "ENSG00000000003    44.46954     28.42961     23.3118      32.25710 \n",
       "ENSG00000000005     0.00000      0.00000      0.0000       0.00000 \n",
       "ENSG00000000419   705.00000    542.00000    787.0000     853.00000 \n",
       "ENSG00000000457   558.08977    519.49518    521.2949     431.25536 \n",
       "ENSG00000000460   117.41435     92.39445     89.9817      98.06563 \n",
       "ENSG00000000938 11941.00000  12390.00000  12951.0000   11913.00000 \n",
       "                3C_S9_R1_001 3D_S10_R1_001 ⋯ 5C_S15_R1_001 5D_S16_R1_001\n",
       "ENSG00000000003    27.76952     33.67071   ⋯    25.53284     16.51529   \n",
       "ENSG00000000005     0.00000      1.00000   ⋯     1.00000      0.00000   \n",
       "ENSG00000000419   814.00000    697.00000   ⋯   683.00000    566.00000   \n",
       "ENSG00000000457   593.09671    780.54047   ⋯   491.88729    469.26460   \n",
       "ENSG00000000460   107.90646    134.14691   ⋯    98.03182     79.98037   \n",
       "ENSG00000000938 14857.00000  10971.00000   ⋯ 11255.00000   9963.00000   \n",
       "                5E_S17_R1_001 5G_S18_R1_001 7A_S19_R1_001 7B_S20_R1_001\n",
       "ENSG00000000003   12.08209       23.44132     18.27322      14.45599   \n",
       "ENSG00000000005    0.00000        0.00000      0.00000       0.00000   \n",
       "ENSG00000000419  523.00000      627.00000    646.00000     561.00000   \n",
       "ENSG00000000457  381.83138      522.13590    474.65664     436.29466   \n",
       "ENSG00000000460   75.98264       95.93890    151.43542      84.70761   \n",
       "ENSG00000000938 9904.00000    10703.00000   8539.00000    9219.00000   \n",
       "                7C_S21_R1_001 7D_S22_R1_001 7E_S23_R1_001 7F_S24_R1_001\n",
       "ENSG00000000003    17.61952      22.95925      37.04031      35.04609  \n",
       "ENSG00000000005     0.00000       0.00000       0.00000       0.00000  \n",
       "ENSG00000000419   702.00000     826.00000     730.00000     755.00000  \n",
       "ENSG00000000457   457.43441     590.91226     538.29130     637.10988  \n",
       "ENSG00000000460   132.63278     164.68933     123.11935     161.07990  \n",
       "ENSG00000000938 10615.00000   12813.00000   14505.00000   13227.00000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(txi$counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 6 × 24 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>1A_S1_R1_001</th><th scope=col>1B_S2_R1_001</th><th scope=col>1C_S3_R1_001</th><th scope=col>1E_S4_R1_001</th><th scope=col>1F_S5_R1_001</th><th scope=col>1G_S6_R1_001</th><th scope=col>3A_S7_R1_001</th><th scope=col>3B_S8_R1_001</th><th scope=col>3C_S9_R1_001</th><th scope=col>3D_S10_R1_001</th><th scope=col>⋯</th><th scope=col>5C_S15_R1_001</th><th scope=col>5D_S16_R1_001</th><th scope=col>5E_S17_R1_001</th><th scope=col>5G_S18_R1_001</th><th scope=col>7A_S19_R1_001</th><th scope=col>7B_S20_R1_001</th><th scope=col>7C_S21_R1_001</th><th scope=col>7D_S22_R1_001</th><th scope=col>7E_S23_R1_001</th><th scope=col>7F_S24_R1_001</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000000003</th><td>  1.462271</td><td>  1.008560</td><td>  1.1787432</td><td>  0.9351069</td><td>  1.365565</td><td>  0.8463663</td><td>  0.5747579</td><td>  0.7650615</td><td>  0.7822711</td><td>  0.9637599</td><td>⋯</td><td>  0.7700359</td><td>  0.6118139</td><td>  0.4458162</td><td>  1.170312</td><td>  0.5818859</td><td>  0.5045505</td><td>  0.5040114</td><td>  0.5271542</td><td>  1.061550</td><td>  0.8454969</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000005</th><td>  0.204246</td><td>  0.000000</td><td>  0.1297491</td><td>  0.0000000</td><td>  0.000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.1015466</td><td>⋯</td><td>  0.1069942</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.0000000</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000419</th><td> 88.474815</td><td> 73.089350</td><td> 73.6780693</td><td> 71.0112082</td><td> 76.143561</td><td> 65.6890103</td><td> 82.0732153</td><td> 81.6783605</td><td> 82.7851247</td><td> 80.7660621</td><td>⋯</td><td> 84.5059962</td><td> 86.3125247</td><td> 77.6555920</td><td> 81.361865</td><td> 83.1550146</td><td> 80.4005526</td><td> 81.1008401</td><td> 76.7301620</td><td> 75.026189</td><td> 72.6233950</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000457</th><td> 16.622367</td><td> 14.598551</td><td> 16.0649069</td><td> 14.1482122</td><td> 14.320505</td><td> 15.4240920</td><td> 11.0171953</td><td>  6.5500682</td><td> 14.2693871</td><td> 23.0335976</td><td>⋯</td><td> 14.0641424</td><td> 14.7664669</td><td> 13.8052737</td><td> 14.925791</td><td> 12.8905529</td><td> 14.3822365</td><td> 13.3374010</td><td> 11.8354678</td><td> 12.181736</td><td> 14.8341505</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000460</th><td>  4.865771</td><td>  5.134641</td><td>  5.3587162</td><td>  3.9200597</td><td>  4.933816</td><td>  2.6953187</td><td>  2.7450779</td><td>  2.8922840</td><td>  4.2197928</td><td>  5.5638859</td><td>⋯</td><td>  4.4160681</td><td>  4.6658248</td><td>  3.5892766</td><td>  3.557216</td><td>  8.0141911</td><td>  3.3057297</td><td>  4.3123775</td><td>  4.7737479</td><td>  4.071993</td><td>  5.3872617</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000938</th><td>531.191987</td><td>614.806153</td><td>486.8702248</td><td>643.5142218</td><td>556.242408</td><td>649.4783794</td><td>561.8686380</td><td>465.7074293</td><td>640.7090237</td><td>558.3422915</td><td>⋯</td><td>584.7271866</td><td>648.2493407</td><td>636.1081486</td><td>610.561735</td><td>467.9885487</td><td>565.5175858</td><td>520.6040472</td><td>516.1287514</td><td>642.586687</td><td>538.0614569</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 24 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & 1A\\_S1\\_R1\\_001 & 1B\\_S2\\_R1\\_001 & 1C\\_S3\\_R1\\_001 & 1E\\_S4\\_R1\\_001 & 1F\\_S5\\_R1\\_001 & 1G\\_S6\\_R1\\_001 & 3A\\_S7\\_R1\\_001 & 3B\\_S8\\_R1\\_001 & 3C\\_S9\\_R1\\_001 & 3D\\_S10\\_R1\\_001 & ⋯ & 5C\\_S15\\_R1\\_001 & 5D\\_S16\\_R1\\_001 & 5E\\_S17\\_R1\\_001 & 5G\\_S18\\_R1\\_001 & 7A\\_S19\\_R1\\_001 & 7B\\_S20\\_R1\\_001 & 7C\\_S21\\_R1\\_001 & 7D\\_S22\\_R1\\_001 & 7E\\_S23\\_R1\\_001 & 7F\\_S24\\_R1\\_001\\\\\n",
       "\\hline\n",
       "\tENSG00000000003 &   1.462271 &   1.008560 &   1.1787432 &   0.9351069 &   1.365565 &   0.8463663 &   0.5747579 &   0.7650615 &   0.7822711 &   0.9637599 & ⋯ &   0.7700359 &   0.6118139 &   0.4458162 &   1.170312 &   0.5818859 &   0.5045505 &   0.5040114 &   0.5271542 &   1.061550 &   0.8454969\\\\\n",
       "\tENSG00000000005 &   0.204246 &   0.000000 &   0.1297491 &   0.0000000 &   0.000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.1015466 & ⋯ &   0.1069942 &   0.0000000 &   0.0000000 &   0.000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.0000000\\\\\n",
       "\tENSG00000000419 &  88.474815 &  73.089350 &  73.6780693 &  71.0112082 &  76.143561 &  65.6890103 &  82.0732153 &  81.6783605 &  82.7851247 &  80.7660621 & ⋯ &  84.5059962 &  86.3125247 &  77.6555920 &  81.361865 &  83.1550146 &  80.4005526 &  81.1008401 &  76.7301620 &  75.026189 &  72.6233950\\\\\n",
       "\tENSG00000000457 &  16.622367 &  14.598551 &  16.0649069 &  14.1482122 &  14.320505 &  15.4240920 &  11.0171953 &   6.5500682 &  14.2693871 &  23.0335976 & ⋯ &  14.0641424 &  14.7664669 &  13.8052737 &  14.925791 &  12.8905529 &  14.3822365 &  13.3374010 &  11.8354678 &  12.181736 &  14.8341505\\\\\n",
       "\tENSG00000000460 &   4.865771 &   5.134641 &   5.3587162 &   3.9200597 &   4.933816 &   2.6953187 &   2.7450779 &   2.8922840 &   4.2197928 &   5.5638859 & ⋯ &   4.4160681 &   4.6658248 &   3.5892766 &   3.557216 &   8.0141911 &   3.3057297 &   4.3123775 &   4.7737479 &   4.071993 &   5.3872617\\\\\n",
       "\tENSG00000000938 & 531.191987 & 614.806153 & 486.8702248 & 643.5142218 & 556.242408 & 649.4783794 & 561.8686380 & 465.7074293 & 640.7090237 & 558.3422915 & ⋯ & 584.7271866 & 648.2493407 & 636.1081486 & 610.561735 & 467.9885487 & 565.5175858 & 520.6040472 & 516.1287514 & 642.586687 & 538.0614569\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 24 of type dbl\n",
       "\n",
       "| <!--/--> | 1A_S1_R1_001 | 1B_S2_R1_001 | 1C_S3_R1_001 | 1E_S4_R1_001 | 1F_S5_R1_001 | 1G_S6_R1_001 | 3A_S7_R1_001 | 3B_S8_R1_001 | 3C_S9_R1_001 | 3D_S10_R1_001 | ⋯ | 5C_S15_R1_001 | 5D_S16_R1_001 | 5E_S17_R1_001 | 5G_S18_R1_001 | 7A_S19_R1_001 | 7B_S20_R1_001 | 7C_S21_R1_001 | 7D_S22_R1_001 | 7E_S23_R1_001 | 7F_S24_R1_001 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ENSG00000000003 |   1.462271 |   1.008560 |   1.1787432 |   0.9351069 |   1.365565 |   0.8463663 |   0.5747579 |   0.7650615 |   0.7822711 |   0.9637599 | ⋯ |   0.7700359 |   0.6118139 |   0.4458162 |   1.170312 |   0.5818859 |   0.5045505 |   0.5040114 |   0.5271542 |   1.061550 |   0.8454969 |\n",
       "| ENSG00000000005 |   0.204246 |   0.000000 |   0.1297491 |   0.0000000 |   0.000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.1015466 | ⋯ |   0.1069942 |   0.0000000 |   0.0000000 |   0.000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.0000000 |\n",
       "| ENSG00000000419 |  88.474815 |  73.089350 |  73.6780693 |  71.0112082 |  76.143561 |  65.6890103 |  82.0732153 |  81.6783605 |  82.7851247 |  80.7660621 | ⋯ |  84.5059962 |  86.3125247 |  77.6555920 |  81.361865 |  83.1550146 |  80.4005526 |  81.1008401 |  76.7301620 |  75.026189 |  72.6233950 |\n",
       "| ENSG00000000457 |  16.622367 |  14.598551 |  16.0649069 |  14.1482122 |  14.320505 |  15.4240920 |  11.0171953 |   6.5500682 |  14.2693871 |  23.0335976 | ⋯ |  14.0641424 |  14.7664669 |  13.8052737 |  14.925791 |  12.8905529 |  14.3822365 |  13.3374010 |  11.8354678 |  12.181736 |  14.8341505 |\n",
       "| ENSG00000000460 |   4.865771 |   5.134641 |   5.3587162 |   3.9200597 |   4.933816 |   2.6953187 |   2.7450779 |   2.8922840 |   4.2197928 |   5.5638859 | ⋯ |   4.4160681 |   4.6658248 |   3.5892766 |   3.557216 |   8.0141911 |   3.3057297 |   4.3123775 |   4.7737479 |   4.071993 |   5.3872617 |\n",
       "| ENSG00000000938 | 531.191987 | 614.806153 | 486.8702248 | 643.5142218 | 556.242408 | 649.4783794 | 561.8686380 | 465.7074293 | 640.7090237 | 558.3422915 | ⋯ | 584.7271866 | 648.2493407 | 636.1081486 | 610.561735 | 467.9885487 | 565.5175858 | 520.6040472 | 516.1287514 | 642.586687 | 538.0614569 |\n",
       "\n"
      ],
      "text/plain": [
       "                1A_S1_R1_001 1B_S2_R1_001 1C_S3_R1_001 1E_S4_R1_001\n",
       "ENSG00000000003   1.462271     1.008560     1.1787432    0.9351069 \n",
       "ENSG00000000005   0.204246     0.000000     0.1297491    0.0000000 \n",
       "ENSG00000000419  88.474815    73.089350    73.6780693   71.0112082 \n",
       "ENSG00000000457  16.622367    14.598551    16.0649069   14.1482122 \n",
       "ENSG00000000460   4.865771     5.134641     5.3587162    3.9200597 \n",
       "ENSG00000000938 531.191987   614.806153   486.8702248  643.5142218 \n",
       "                1F_S5_R1_001 1G_S6_R1_001 3A_S7_R1_001 3B_S8_R1_001\n",
       "ENSG00000000003   1.365565     0.8463663    0.5747579    0.7650615 \n",
       "ENSG00000000005   0.000000     0.0000000    0.0000000    0.0000000 \n",
       "ENSG00000000419  76.143561    65.6890103   82.0732153   81.6783605 \n",
       "ENSG00000000457  14.320505    15.4240920   11.0171953    6.5500682 \n",
       "ENSG00000000460   4.933816     2.6953187    2.7450779    2.8922840 \n",
       "ENSG00000000938 556.242408   649.4783794  561.8686380  465.7074293 \n",
       "                3C_S9_R1_001 3D_S10_R1_001 ⋯ 5C_S15_R1_001 5D_S16_R1_001\n",
       "ENSG00000000003   0.7822711    0.9637599   ⋯   0.7700359     0.6118139  \n",
       "ENSG00000000005   0.0000000    0.1015466   ⋯   0.1069942     0.0000000  \n",
       "ENSG00000000419  82.7851247   80.7660621   ⋯  84.5059962    86.3125247  \n",
       "ENSG00000000457  14.2693871   23.0335976   ⋯  14.0641424    14.7664669  \n",
       "ENSG00000000460   4.2197928    5.5638859   ⋯   4.4160681     4.6658248  \n",
       "ENSG00000000938 640.7090237  558.3422915   ⋯ 584.7271866   648.2493407  \n",
       "                5E_S17_R1_001 5G_S18_R1_001 7A_S19_R1_001 7B_S20_R1_001\n",
       "ENSG00000000003   0.4458162     1.170312      0.5818859     0.5045505  \n",
       "ENSG00000000005   0.0000000     0.000000      0.0000000     0.0000000  \n",
       "ENSG00000000419  77.6555920    81.361865     83.1550146    80.4005526  \n",
       "ENSG00000000457  13.8052737    14.925791     12.8905529    14.3822365  \n",
       "ENSG00000000460   3.5892766     3.557216      8.0141911     3.3057297  \n",
       "ENSG00000000938 636.1081486   610.561735    467.9885487   565.5175858  \n",
       "                7C_S21_R1_001 7D_S22_R1_001 7E_S23_R1_001 7F_S24_R1_001\n",
       "ENSG00000000003   0.5040114     0.5271542     1.061550      0.8454969  \n",
       "ENSG00000000005   0.0000000     0.0000000     0.000000      0.0000000  \n",
       "ENSG00000000419  81.1008401    76.7301620    75.026189     72.6233950  \n",
       "ENSG00000000457  13.3374010    11.8354678    12.181736     14.8341505  \n",
       "ENSG00000000460   4.3123775     4.7737479     4.071993      5.3872617  \n",
       "ENSG00000000938 520.6040472   516.1287514   642.586687    538.0614569  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(txi$abundance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'abundance'</li><li>'counts'</li><li>'length'</li><li>'countsFromAbundance'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'abundance'\n",
       "\\item 'counts'\n",
       "\\item 'length'\n",
       "\\item 'countsFromAbundance'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'abundance'\n",
       "2. 'counts'\n",
       "3. 'length'\n",
       "4. 'countsFromAbundance'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"abundance\"           \"counts\"              \"length\"             \n",
       "[4] \"countsFromAbundance\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "names(txi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for tximport {tximport}\"><tr><td>tximport {tximport}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Import transcript-level abundances and counts\n",
       "for transcript- and gene-level analysis packages</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p><code>tximport</code> imports transcript-level estimates from various\n",
       "external software and optionally summarizes abundances, counts,\n",
       "and transcript lengths\n",
       "to the gene-level (default) or outputs transcript-level matrices\n",
       "(see <code>txOut</code> argument).\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "tximport(\n",
       "  files,\n",
       "  type = c(\"none\", \"salmon\", \"sailfish\", \"alevin\", \"kallisto\", \"rsem\", \"stringtie\"),\n",
       "  txIn = TRUE,\n",
       "  txOut = FALSE,\n",
       "  countsFromAbundance = c(\"no\", \"scaledTPM\", \"lengthScaledTPM\", \"dtuScaledTPM\"),\n",
       "  tx2gene = NULL,\n",
       "  varReduce = FALSE,\n",
       "  dropInfReps = FALSE,\n",
       "  infRepStat = NULL,\n",
       "  ignoreTxVersion = FALSE,\n",
       "  ignoreAfterBar = FALSE,\n",
       "  geneIdCol,\n",
       "  txIdCol,\n",
       "  abundanceCol,\n",
       "  countsCol,\n",
       "  lengthCol,\n",
       "  importer = NULL,\n",
       "  existenceOptional = FALSE,\n",
       "  sparse = FALSE,\n",
       "  sparseThreshold = 1,\n",
       "  readLength = 75,\n",
       "  alevinArgs = NULL\n",
       ")\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>files</code></td>\n",
       "<td>\n",
       "<p>a character vector of filenames for the transcript-level abundances</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>type</code></td>\n",
       "<td>\n",
       "<p>character, the type of software used to generate the abundances.\n",
       "Options are &quot;salmon&quot;, &quot;sailfish&quot;, &quot;alevin&quot;, &quot;kallisto&quot;, &quot;rsem&quot;, &quot;stringtie&quot;, or &quot;none&quot;.\n",
       "This argument is used to autofill the arguments below (geneIdCol, etc.)\n",
       "&quot;none&quot; means that the user will specify these columns.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>txIn</code></td>\n",
       "<td>\n",
       "<p>logical, whether the incoming files are transcript level (default TRUE)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>txOut</code></td>\n",
       "<td>\n",
       "<p>logical, whether the function should just output\n",
       "transcript-level (default FALSE)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>countsFromAbundance</code></td>\n",
       "<td>\n",
       "<p>character, either &quot;no&quot; (default), &quot;scaledTPM&quot;,\n",
       "&quot;lengthScaledTPM&quot;, or &quot;dtuScaledTPM&quot;.\n",
       "Whether to generate estimated counts using abundance estimates:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p> scaled up to library size (scaledTPM),\n",
       "</p>\n",
       "</li>\n",
       "<li><p> scaled using the average transcript length over samples\n",
       "and then the library size (lengthScaledTPM), or\n",
       "</p>\n",
       "</li>\n",
       "<li><p> scaled using the median transcript length among isoforms of a gene,\n",
       "and then the library size (dtuScaledTPM). \n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "<p>dtuScaledTPM is designed for DTU analysis in combination with <code>txOut=TRUE</code>,\n",
       "and it requires specifing a <code>tx2gene</code> data.frame.\n",
       "dtuScaledTPM works such that within a gene, values from all samples and\n",
       "all transcripts get scaled by the same fixed median transcript length.\n",
       "If using scaledTPM, lengthScaledTPM, or geneLengthScaledTPM, \n",
       "the counts are no longer correlated across samples with transcript length,\n",
       "and so the length offset matrix should not be used.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>tx2gene</code></td>\n",
       "<td>\n",
       "<p>a two-column data.frame linking transcript id (column 1)\n",
       "to gene id (column 2).\n",
       "the column names are not relevant, but this column order must be used. \n",
       "this argument is required for gene-level summarization, and the tximport\n",
       "vignette describes how to construct this data.frame (see Details below).\n",
       "An automated solution to avoid having to create <code>tx2gene</code> if\n",
       "one has quantified with Salmon or alevin with human or mouse transcriptomes\n",
       "is to use the <code>tximeta</code> function from the tximeta Bioconductor package.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>varReduce</code></td>\n",
       "<td>\n",
       "<p>whether to reduce per-sample inferential replicates\n",
       "information into a matrix of sample variances <code>variance</code> (default FALSE).\n",
       "alevin computes inferential variance by default for bootstrap\n",
       "inferential replicates, so this argument is ignored/not necessary</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>dropInfReps</code></td>\n",
       "<td>\n",
       "<p>whether to skip reading in inferential replicates\n",
       "(default FALSE). For alevin, <code>tximport</code> will still read in the\n",
       "inferential variance matrix if it exists</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>infRepStat</code></td>\n",
       "<td>\n",
       "<p>a function to re-compute counts and abundances from the\n",
       "inferential replicates, e.g. <code>matrixStats::rowMedians</code> to re-compute counts \n",
       "as the median of the inferential replicates. The order of operations is:\n",
       "first counts are re-computed, then abundances are re-computed.\n",
       "Following this, if <code>countsFromAbundance</code> is not &quot;no&quot;,\n",
       "<code>tximport</code> will again re-compute counts from the re-computed abundances.\n",
       "<code>infRepStat</code> should operate on rows of a matrix. (default is NULL)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>ignoreTxVersion</code></td>\n",
       "<td>\n",
       "<p>logical, whether to split the tx id on the '.' character\n",
       "to remove version information to facilitate matching with the tx id in <code>tx2gene</code>\n",
       "(default FALSE)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>ignoreAfterBar</code></td>\n",
       "<td>\n",
       "<p>logical, whether to split the tx id on the '|' character\n",
       "to facilitate matching with the tx id in <code>tx2gene</code> (default FALSE)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>geneIdCol</code></td>\n",
       "<td>\n",
       "<p>name of column with gene id. if missing, the <code>tx2gene</code>\n",
       "argument can be used</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>txIdCol</code></td>\n",
       "<td>\n",
       "<p>name of column with tx id</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>abundanceCol</code></td>\n",
       "<td>\n",
       "<p>name of column with abundances (e.g. TPM or FPKM)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>countsCol</code></td>\n",
       "<td>\n",
       "<p>name of column with estimated counts</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>lengthCol</code></td>\n",
       "<td>\n",
       "<p>name of column with feature length information</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>importer</code></td>\n",
       "<td>\n",
       "<p>a function used to read in the files</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>existenceOptional</code></td>\n",
       "<td>\n",
       "<p>logical, should tximport not check if files exist before attempting\n",
       "import (default FALSE, meaning files must exist according to <code>file.exists</code>)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>sparse</code></td>\n",
       "<td>\n",
       "<p>logical, whether to try to import data sparsely (default is FALSE).\n",
       "Initial implementation for <code>txOut=TRUE</code>, <code>countsFromAbundance=\"no\"</code>\n",
       "or <code>\"scaledTPM\"</code>, no inferential replicates. Only counts matrix\n",
       "is returned (and abundance matrix if using <code>\"scaledTPM\"</code>)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>sparseThreshold</code></td>\n",
       "<td>\n",
       "<p>the minimum threshold for including a count as a\n",
       "non-zero count during sparse import (default is 1)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>readLength</code></td>\n",
       "<td>\n",
       "<p>numeric, the read length used to calculate counts from\n",
       "StringTie's output of coverage. Default value (from StringTie) is 75.\n",
       "The formula used to calculate counts is:\n",
       "<code>cov * transcript length / read length</code></p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>alevinArgs</code></td>\n",
       "<td>\n",
       "<p>named list, with logical elements <code>filterBarcodes</code>,\n",
       "<code>tierImport</code>, <code>forceSlow</code>. See Details for definitions.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p><strong>Inferential replicates:</strong>\n",
       "<code>tximport</code> will also load in information about inferential replicates &ndash;\n",
       "a list of matrices of the Gibbs samples from the posterior, or bootstrap replicates,\n",
       "per sample &ndash; if these data are available in the expected locations relative\n",
       "to the <code>files</code>.\n",
       "The inferential replicates, stored in <code>infReps</code> in the output list,\n",
       "are on estimated counts, and therefore follow <code>counts</code> in the output list.\n",
       "By setting <code>varReduce=TRUE</code>, the inferential replicate matrices\n",
       "will be replaced by a single matrix with the sample variance per transcript/gene\n",
       "and per sample.\n",
       "</p>\n",
       "<p><strong>summarizeToGene:</strong>\n",
       "While <code>tximport</code> summarizes to the gene-level by default, \n",
       "the user can also perform the import and summarization steps manually,\n",
       "by specifing <code>txOut=TRUE</code> and then using the function <code>summarizeToGene</code>.\n",
       "Note however that this is equivalent to <code>tximport</code> with\n",
       "<code>txOut=FALSE</code> (the default).\n",
       "</p>\n",
       "<p><strong>Solutions on summarization:</strong> regarding <code>\"tximport failed at summarizing to the gene-level\"</code>:\n",
       "</p>\n",
       "\n",
       "<ol>\n",
       "<li><p> provide a <code>tx2gene</code> data.frame linking transcripts to genes (more below)\n",
       "</p>\n",
       "</li>\n",
       "<li><p> avoid gene-level summarization by specifying <code>txOut=TRUE</code>\n",
       "</p>\n",
       "</li></ol>\n",
       "\n",
       "<p>See <code>vignette('tximport')</code> for example code for generating a\n",
       "<code>tx2gene</code> data.frame from a <code>TxDb</code> object.\n",
       "The <code>tx2gene</code> data.frame should exactly match and be derived from\n",
       "the same set of transcripts used for quantifying (the set of transcript\n",
       "used to create the transcriptome index).\n",
       "</p>\n",
       "<p><strong>Tximeta:</strong>\n",
       "One automated solution for Salmon or alevin quantification data is to use the\n",
       "<code>tximeta</code> function in the tximeta Bioconductor package\n",
       "which builds upon and extends <code>tximport</code>; this solution should\n",
       "work out-of-the-box for human and mouse transcriptomes downloaded\n",
       "from GENCODE, Ensembl, or RefSeq. For other cases, the user\n",
       "should create the <code>tx2gene</code> manually as shown in the tximport\n",
       "vignette.\n",
       "</p>\n",
       "<p><strong>On tx2gene construction:</strong>\n",
       "Note that the <code>keys</code> and <code>select</code> functions used\n",
       "to create the <code>tx2gene</code> object are documented\n",
       "in the man page for AnnotationDb-class objects\n",
       "in the AnnotationDbi package (TxDb inherits from AnnotationDb).\n",
       "For further details on generating TxDb objects from various inputs\n",
       "see <code>vignette('GenomicFeatures')</code> from the GenomicFeatures package.\n",
       "</p>\n",
       "<p><strong>alevin:</strong>\n",
       "The <code>alevinArgs</code> argument includes some alevin-specific arguments.\n",
       "This optional argument is a list with any or all of the following named logical variables:\n",
       "<code>filterBarcodes</code>, <code>tierImport</code>, and <code>forceSlow</code>.\n",
       "The variables are described as follows (with default values in parens):\n",
       "<code>filterBarcodes</code> (FALSE) import only cell barcodes listed in\n",
       "<code>whitelist.txt</code>;\n",
       "<code>tierImport</code> (FALSE) import the tier information in addition to counts;\n",
       "<code>forceSlow</code> (FALSE) force the use of the slower import R code\n",
       "even if <code>fishpond</code> is installed.\n",
       "For <code>type=\"alevin\"</code> all arguments other than <code>files</code>,\n",
       "<code>dropInfReps</code>, and <code>alevinArgs</code> are ignored.\n",
       "Note that <code>files</code> should point to a single <code>quants_mat.gz</code> file,\n",
       "in the directory structure created by the alevin software\n",
       "(e.g. do not move the file or delete the other important files).\n",
       "Note that importing alevin quantifications will be much faster by first\n",
       "installing the <code>fishpond</code> package, which contains a C++ importer\n",
       "for alevin's EDS format.\n",
       "For alevin, <code>tximport</code> is importing the gene-by-cell matrix of counts,\n",
       "as <code>txi$counts</code>, and effective lengths are not estimated.\n",
       "<code>txi$mean</code> and <code>txi$variance</code> may also be imported if\n",
       "inferential replicates were used, as well as inferential replicates\n",
       "if these were output by alevin.\n",
       "Length correction should not be applied to datasets where there\n",
       "is not an expected correlation of counts and feature length.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>A simple list containing matrices: abundance, counts, length.\n",
       "Another list element 'countsFromAbundance' carries through\n",
       "the character argument used in the tximport call.\n",
       "The length matrix contains the average transcript length for each\n",
       "gene which can be used as an offset for gene-level analysis.\n",
       "If detected, and <code>txOut=TRUE</code>, inferential replicates for\n",
       "each sample will be imported and stored as a list of matrices,\n",
       "itself an element <code>infReps</code> in the returned list.\n",
       "An exception is alevin, in which the <code>infReps</code> are a list\n",
       "of bootstrap replicate matrices, where each matrix has\n",
       "genes as rows and cells as columns.\n",
       "If <code>varReduce=TRUE</code> the inferential replicates will be summarized\n",
       "according to the sample variance, and stored as a matrix <code>variance</code>.\n",
       "alevin already computes the variance of the bootstrap inferential replicates\n",
       "and so this is imported without needing to specify <code>varReduce=TRUE</code>\n",
       "(note that alevin uses the 1/N variance estimator, so not the same as <code>var</code>).\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>Charlotte Soneson, Michael I. Love, Mark D. Robinson (2015)\n",
       "Differential analyses for RNA-seq: transcript-level estimates\n",
       "improve gene-level inferences. F1000Research.\n",
       "<a href=\"http://doi.org/10.12688/f1000research.7563\">http://doi.org/10.12688/f1000research.7563</a>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "\n",
       "# load data for demonstrating tximport\n",
       "# note that the vignette shows more examples\n",
       "# including how to read in files quickly using the readr package\n",
       "\n",
       "library(tximportData)\n",
       "dir &lt;- system.file(\"extdata\", package=\"tximportData\")\n",
       "samples &lt;- read.table(file.path(dir,\"samples.txt\"), header=TRUE)\n",
       "files &lt;- file.path(dir,\"salmon\", samples$run, \"quant.sf.gz\")\n",
       "names(files) &lt;- paste0(\"sample\",1:6)\n",
       "\n",
       "# tx2gene links transcript IDs to gene IDs for summarization\n",
       "tx2gene &lt;- read.csv(file.path(dir, \"tx2gene.gencode.v27.csv\"))\n",
       "\n",
       "txi &lt;- tximport(files, type=\"salmon\", tx2gene=tx2gene)\n",
       "\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>tximport</em> version 1.16.1 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{tximport}{Import transcript-level abundances and counts for transcript- and gene-level analysis packages}{tximport}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "\\code{tximport} imports transcript-level estimates from various\n",
       "external software and optionally summarizes abundances, counts,\n",
       "and transcript lengths\n",
       "to the gene-level (default) or outputs transcript-level matrices\n",
       "(see \\code{txOut} argument).\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "tximport(\n",
       "  files,\n",
       "  type = c(\"none\", \"salmon\", \"sailfish\", \"alevin\", \"kallisto\", \"rsem\", \"stringtie\"),\n",
       "  txIn = TRUE,\n",
       "  txOut = FALSE,\n",
       "  countsFromAbundance = c(\"no\", \"scaledTPM\", \"lengthScaledTPM\", \"dtuScaledTPM\"),\n",
       "  tx2gene = NULL,\n",
       "  varReduce = FALSE,\n",
       "  dropInfReps = FALSE,\n",
       "  infRepStat = NULL,\n",
       "  ignoreTxVersion = FALSE,\n",
       "  ignoreAfterBar = FALSE,\n",
       "  geneIdCol,\n",
       "  txIdCol,\n",
       "  abundanceCol,\n",
       "  countsCol,\n",
       "  lengthCol,\n",
       "  importer = NULL,\n",
       "  existenceOptional = FALSE,\n",
       "  sparse = FALSE,\n",
       "  sparseThreshold = 1,\n",
       "  readLength = 75,\n",
       "  alevinArgs = NULL\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{files}] a character vector of filenames for the transcript-level abundances\n",
       "\n",
       "\\item[\\code{type}] character, the type of software used to generate the abundances.\n",
       "Options are \"salmon\", \"sailfish\", \"alevin\", \"kallisto\", \"rsem\", \"stringtie\", or \"none\".\n",
       "This argument is used to autofill the arguments below (geneIdCol, etc.)\n",
       "\"none\" means that the user will specify these columns.\n",
       "\n",
       "\\item[\\code{txIn}] logical, whether the incoming files are transcript level (default TRUE)\n",
       "\n",
       "\\item[\\code{txOut}] logical, whether the function should just output\n",
       "transcript-level (default FALSE)\n",
       "\n",
       "\\item[\\code{countsFromAbundance}] character, either \"no\" (default), \"scaledTPM\",\n",
       "\"lengthScaledTPM\", or \"dtuScaledTPM\".\n",
       "Whether to generate estimated counts using abundance estimates:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} scaled up to library size (scaledTPM),\n",
       "\\item{} scaled using the average transcript length over samples\n",
       "and then the library size (lengthScaledTPM), or\n",
       "\\item{} scaled using the median transcript length among isoforms of a gene,\n",
       "and then the library size (dtuScaledTPM). \n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "dtuScaledTPM is designed for DTU analysis in combination with \\code{txOut=TRUE},\n",
       "and it requires specifing a \\code{tx2gene} data.frame.\n",
       "dtuScaledTPM works such that within a gene, values from all samples and\n",
       "all transcripts get scaled by the same fixed median transcript length.\n",
       "If using scaledTPM, lengthScaledTPM, or geneLengthScaledTPM, \n",
       "the counts are no longer correlated across samples with transcript length,\n",
       "and so the length offset matrix should not be used.\n",
       "\n",
       "\\item[\\code{tx2gene}] a two-column data.frame linking transcript id (column 1)\n",
       "to gene id (column 2).\n",
       "the column names are not relevant, but this column order must be used. \n",
       "this argument is required for gene-level summarization, and the tximport\n",
       "vignette describes how to construct this data.frame (see Details below).\n",
       "An automated solution to avoid having to create \\code{tx2gene} if\n",
       "one has quantified with Salmon or alevin with human or mouse transcriptomes\n",
       "is to use the \\code{tximeta} function from the tximeta Bioconductor package.\n",
       "\n",
       "\\item[\\code{varReduce}] whether to reduce per-sample inferential replicates\n",
       "information into a matrix of sample variances \\code{variance} (default FALSE).\n",
       "alevin computes inferential variance by default for bootstrap\n",
       "inferential replicates, so this argument is ignored/not necessary\n",
       "\n",
       "\\item[\\code{dropInfReps}] whether to skip reading in inferential replicates\n",
       "(default FALSE). For alevin, \\code{tximport} will still read in the\n",
       "inferential variance matrix if it exists\n",
       "\n",
       "\\item[\\code{infRepStat}] a function to re-compute counts and abundances from the\n",
       "inferential replicates, e.g. \\code{matrixStats::rowMedians} to re-compute counts \n",
       "as the median of the inferential replicates. The order of operations is:\n",
       "first counts are re-computed, then abundances are re-computed.\n",
       "Following this, if \\code{countsFromAbundance} is not \"no\",\n",
       "\\code{tximport} will again re-compute counts from the re-computed abundances.\n",
       "\\code{infRepStat} should operate on rows of a matrix. (default is NULL)\n",
       "\n",
       "\\item[\\code{ignoreTxVersion}] logical, whether to split the tx id on the '.' character\n",
       "to remove version information to facilitate matching with the tx id in \\code{tx2gene}\n",
       "(default FALSE)\n",
       "\n",
       "\\item[\\code{ignoreAfterBar}] logical, whether to split the tx id on the '|' character\n",
       "to facilitate matching with the tx id in \\code{tx2gene} (default FALSE)\n",
       "\n",
       "\\item[\\code{geneIdCol}] name of column with gene id. if missing, the \\code{tx2gene}\n",
       "argument can be used\n",
       "\n",
       "\\item[\\code{txIdCol}] name of column with tx id\n",
       "\n",
       "\\item[\\code{abundanceCol}] name of column with abundances (e.g. TPM or FPKM)\n",
       "\n",
       "\\item[\\code{countsCol}] name of column with estimated counts\n",
       "\n",
       "\\item[\\code{lengthCol}] name of column with feature length information\n",
       "\n",
       "\\item[\\code{importer}] a function used to read in the files\n",
       "\n",
       "\\item[\\code{existenceOptional}] logical, should tximport not check if files exist before attempting\n",
       "import (default FALSE, meaning files must exist according to \\code{file.exists})\n",
       "\n",
       "\\item[\\code{sparse}] logical, whether to try to import data sparsely (default is FALSE).\n",
       "Initial implementation for \\code{txOut=TRUE}, \\code{countsFromAbundance=\"no\"}\n",
       "or \\code{\"scaledTPM\"}, no inferential replicates. Only counts matrix\n",
       "is returned (and abundance matrix if using \\code{\"scaledTPM\"})\n",
       "\n",
       "\\item[\\code{sparseThreshold}] the minimum threshold for including a count as a\n",
       "non-zero count during sparse import (default is 1)\n",
       "\n",
       "\\item[\\code{readLength}] numeric, the read length used to calculate counts from\n",
       "StringTie's output of coverage. Default value (from StringTie) is 75.\n",
       "The formula used to calculate counts is:\n",
       "\\code{cov * transcript length / read length}\n",
       "\n",
       "\\item[\\code{alevinArgs}] named list, with logical elements \\code{filterBarcodes},\n",
       "\\code{tierImport}, \\code{forceSlow}. See Details for definitions.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "\\strong{Inferential replicates:}\n",
       "\\code{tximport} will also load in information about inferential replicates --\n",
       "a list of matrices of the Gibbs samples from the posterior, or bootstrap replicates,\n",
       "per sample -- if these data are available in the expected locations relative\n",
       "to the \\code{files}.\n",
       "The inferential replicates, stored in \\code{infReps} in the output list,\n",
       "are on estimated counts, and therefore follow \\code{counts} in the output list.\n",
       "By setting \\code{varReduce=TRUE}, the inferential replicate matrices\n",
       "will be replaced by a single matrix with the sample variance per transcript/gene\n",
       "and per sample.\n",
       "\n",
       "\\strong{summarizeToGene:}\n",
       "While \\code{tximport} summarizes to the gene-level by default, \n",
       "the user can also perform the import and summarization steps manually,\n",
       "by specifing \\code{txOut=TRUE} and then using the function \\code{summarizeToGene}.\n",
       "Note however that this is equivalent to \\code{tximport} with\n",
       "\\code{txOut=FALSE} (the default).\n",
       "\n",
       "\\strong{Solutions on summarization:} regarding \\code{\"tximport failed at summarizing to the gene-level\"}:\n",
       "\n",
       "\\begin{enumerate}\n",
       "\n",
       "\\item{} provide a \\code{tx2gene} data.frame linking transcripts to genes (more below)\n",
       "\\item{} avoid gene-level summarization by specifying \\code{txOut=TRUE}\n",
       "\n",
       "\\end{enumerate}\n",
       "\n",
       "\n",
       "See \\code{vignette('tximport')} for example code for generating a\n",
       "\\code{tx2gene} data.frame from a \\code{TxDb} object.\n",
       "The \\code{tx2gene} data.frame should exactly match and be derived from\n",
       "the same set of transcripts used for quantifying (the set of transcript\n",
       "used to create the transcriptome index).\n",
       "\n",
       "\\strong{Tximeta:}\n",
       "One automated solution for Salmon or alevin quantification data is to use the\n",
       "\\code{tximeta} function in the tximeta Bioconductor package\n",
       "which builds upon and extends \\code{tximport}; this solution should\n",
       "work out-of-the-box for human and mouse transcriptomes downloaded\n",
       "from GENCODE, Ensembl, or RefSeq. For other cases, the user\n",
       "should create the \\code{tx2gene} manually as shown in the tximport\n",
       "vignette.\n",
       "\n",
       "\\strong{On tx2gene construction:}\n",
       "Note that the \\code{keys} and \\code{select} functions used\n",
       "to create the \\code{tx2gene} object are documented\n",
       "in the man page for \\LinkA{AnnotationDb-class}{AnnotationDb.Rdash.class} objects\n",
       "in the AnnotationDbi package (TxDb inherits from AnnotationDb).\n",
       "For further details on generating TxDb objects from various inputs\n",
       "see \\code{vignette('GenomicFeatures')} from the GenomicFeatures package.\n",
       "\n",
       "\\strong{alevin:}\n",
       "The \\code{alevinArgs} argument includes some alevin-specific arguments.\n",
       "This optional argument is a list with any or all of the following named logical variables:\n",
       "\\code{filterBarcodes}, \\code{tierImport}, and \\code{forceSlow}.\n",
       "The variables are described as follows (with default values in parens):\n",
       "\\code{filterBarcodes} (FALSE) import only cell barcodes listed in\n",
       "\\code{whitelist.txt};\n",
       "\\code{tierImport} (FALSE) import the tier information in addition to counts;\n",
       "\\code{forceSlow} (FALSE) force the use of the slower import R code\n",
       "even if \\code{fishpond} is installed.\n",
       "For \\code{type=\"alevin\"} all arguments other than \\code{files},\n",
       "\\code{dropInfReps}, and \\code{alevinArgs} are ignored.\n",
       "Note that \\code{files} should point to a single \\code{quants\\_mat.gz} file,\n",
       "in the directory structure created by the alevin software\n",
       "(e.g. do not move the file or delete the other important files).\n",
       "Note that importing alevin quantifications will be much faster by first\n",
       "installing the \\code{fishpond} package, which contains a C++ importer\n",
       "for alevin's EDS format.\n",
       "For alevin, \\code{tximport} is importing the gene-by-cell matrix of counts,\n",
       "as \\code{txi\\$counts}, and effective lengths are not estimated.\n",
       "\\code{txi\\$mean} and \\code{txi\\$variance} may also be imported if\n",
       "inferential replicates were used, as well as inferential replicates\n",
       "if these were output by alevin.\n",
       "Length correction should not be applied to datasets where there\n",
       "is not an expected correlation of counts and feature length.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "A simple list containing matrices: abundance, counts, length.\n",
       "Another list element 'countsFromAbundance' carries through\n",
       "the character argument used in the tximport call.\n",
       "The length matrix contains the average transcript length for each\n",
       "gene which can be used as an offset for gene-level analysis.\n",
       "If detected, and \\code{txOut=TRUE}, inferential replicates for\n",
       "each sample will be imported and stored as a list of matrices,\n",
       "itself an element \\code{infReps} in the returned list.\n",
       "An exception is alevin, in which the \\code{infReps} are a list\n",
       "of bootstrap replicate matrices, where each matrix has\n",
       "genes as rows and cells as columns.\n",
       "If \\code{varReduce=TRUE} the inferential replicates will be summarized\n",
       "according to the sample variance, and stored as a matrix \\code{variance}.\n",
       "alevin already computes the variance of the bootstrap inferential replicates\n",
       "and so this is imported without needing to specify \\code{varReduce=TRUE}\n",
       "(note that alevin uses the 1/N variance estimator, so not the same as \\code{var}).\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "Charlotte Soneson, Michael I. Love, Mark D. Robinson (2015)\n",
       "Differential analyses for RNA-seq: transcript-level estimates\n",
       "improve gene-level inferences. F1000Research.\n",
       "\\url{http://doi.org/10.12688/f1000research.7563}\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "\n",
       "# load data for demonstrating tximport\n",
       "# note that the vignette shows more examples\n",
       "# including how to read in files quickly using the readr package\n",
       "\n",
       "library(tximportData)\n",
       "dir <- system.file(\"extdata\", package=\"tximportData\")\n",
       "samples <- read.table(file.path(dir,\"samples.txt\"), header=TRUE)\n",
       "files <- file.path(dir,\"salmon\", samples$run, \"quant.sf.gz\")\n",
       "names(files) <- paste0(\"sample\",1:6)\n",
       "\n",
       "# tx2gene links transcript IDs to gene IDs for summarization\n",
       "tx2gene <- read.csv(file.path(dir, \"tx2gene.gencode.v27.csv\"))\n",
       "\n",
       "txi <- tximport(files, type=\"salmon\", tx2gene=tx2gene)\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "tximport               package:tximport                R Documentation\n",
       "\n",
       "_\bI_\bm_\bp_\bo_\br_\bt _\bt_\br_\ba_\bn_\bs_\bc_\br_\bi_\bp_\bt-_\bl_\be_\bv_\be_\bl _\ba_\bb_\bu_\bn_\bd_\ba_\bn_\bc_\be_\bs _\ba_\bn_\bd _\bc_\bo_\bu_\bn_\bt_\bs _\bf_\bo_\br _\bt_\br_\ba_\bn_\bs_\bc_\br_\bi_\bp_\bt- _\ba_\bn_\bd\n",
       "_\bg_\be_\bn_\be-_\bl_\be_\bv_\be_\bl _\ba_\bn_\ba_\bl_\by_\bs_\bi_\bs _\bp_\ba_\bc_\bk_\ba_\bg_\be_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     ‘tximport’ imports transcript-level estimates from various\n",
       "     external software and optionally summarizes abundances, counts,\n",
       "     and transcript lengths to the gene-level (default) or outputs\n",
       "     transcript-level matrices (see ‘txOut’ argument).\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     tximport(\n",
       "       files,\n",
       "       type = c(\"none\", \"salmon\", \"sailfish\", \"alevin\", \"kallisto\", \"rsem\", \"stringtie\"),\n",
       "       txIn = TRUE,\n",
       "       txOut = FALSE,\n",
       "       countsFromAbundance = c(\"no\", \"scaledTPM\", \"lengthScaledTPM\", \"dtuScaledTPM\"),\n",
       "       tx2gene = NULL,\n",
       "       varReduce = FALSE,\n",
       "       dropInfReps = FALSE,\n",
       "       infRepStat = NULL,\n",
       "       ignoreTxVersion = FALSE,\n",
       "       ignoreAfterBar = FALSE,\n",
       "       geneIdCol,\n",
       "       txIdCol,\n",
       "       abundanceCol,\n",
       "       countsCol,\n",
       "       lengthCol,\n",
       "       importer = NULL,\n",
       "       existenceOptional = FALSE,\n",
       "       sparse = FALSE,\n",
       "       sparseThreshold = 1,\n",
       "       readLength = 75,\n",
       "       alevinArgs = NULL\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "   files: a character vector of filenames for the transcript-level\n",
       "          abundances\n",
       "\n",
       "    type: character, the type of software used to generate the\n",
       "          abundances. Options are \"salmon\", \"sailfish\", \"alevin\",\n",
       "          \"kallisto\", \"rsem\", \"stringtie\", or \"none\". This argument is\n",
       "          used to autofill the arguments below (geneIdCol, etc.) \"none\"\n",
       "          means that the user will specify these columns.\n",
       "\n",
       "    txIn: logical, whether the incoming files are transcript level\n",
       "          (default TRUE)\n",
       "\n",
       "   txOut: logical, whether the function should just output\n",
       "          transcript-level (default FALSE)\n",
       "\n",
       "countsFromAbundance: character, either \"no\" (default), \"scaledTPM\",\n",
       "          \"lengthScaledTPM\", or \"dtuScaledTPM\". Whether to generate\n",
       "          estimated counts using abundance estimates:\n",
       "\n",
       "            • scaled up to library size (scaledTPM),\n",
       "\n",
       "            • scaled using the average transcript length over samples\n",
       "              and then the library size (lengthScaledTPM), or\n",
       "\n",
       "            • scaled using the median transcript length among isoforms\n",
       "              of a gene, and then the library size (dtuScaledTPM).\n",
       "\n",
       "          dtuScaledTPM is designed for DTU analysis in combination with\n",
       "          ‘txOut=TRUE’, and it requires specifing a ‘tx2gene’\n",
       "          data.frame. dtuScaledTPM works such that within a gene,\n",
       "          values from all samples and all transcripts get scaled by the\n",
       "          same fixed median transcript length. If using scaledTPM,\n",
       "          lengthScaledTPM, or geneLengthScaledTPM, the counts are no\n",
       "          longer correlated across samples with transcript length, and\n",
       "          so the length offset matrix should not be used.\n",
       "\n",
       " tx2gene: a two-column data.frame linking transcript id (column 1) to\n",
       "          gene id (column 2). the column names are not relevant, but\n",
       "          this column order must be used.  this argument is required\n",
       "          for gene-level summarization, and the tximport vignette\n",
       "          describes how to construct this data.frame (see Details\n",
       "          below). An automated solution to avoid having to create\n",
       "          ‘tx2gene’ if one has quantified with Salmon or alevin with\n",
       "          human or mouse transcriptomes is to use the ‘tximeta’\n",
       "          function from the tximeta Bioconductor package.\n",
       "\n",
       "varReduce: whether to reduce per-sample inferential replicates\n",
       "          information into a matrix of sample variances ‘variance’\n",
       "          (default FALSE). alevin computes inferential variance by\n",
       "          default for bootstrap inferential replicates, so this\n",
       "          argument is ignored/not necessary\n",
       "\n",
       "dropInfReps: whether to skip reading in inferential replicates (default\n",
       "          FALSE). For alevin, ‘tximport’ will still read in the\n",
       "          inferential variance matrix if it exists\n",
       "\n",
       "infRepStat: a function to re-compute counts and abundances from the\n",
       "          inferential replicates, e.g. ‘matrixStats::rowMedians’ to\n",
       "          re-compute counts as the median of the inferential\n",
       "          replicates. The order of operations is: first counts are\n",
       "          re-computed, then abundances are re-computed. Following this,\n",
       "          if ‘countsFromAbundance’ is not \"no\", ‘tximport’ will again\n",
       "          re-compute counts from the re-computed abundances.\n",
       "          ‘infRepStat’ should operate on rows of a matrix. (default is\n",
       "          NULL)\n",
       "\n",
       "ignoreTxVersion: logical, whether to split the tx id on the '.'\n",
       "          character to remove version information to facilitate\n",
       "          matching with the tx id in ‘tx2gene’ (default FALSE)\n",
       "\n",
       "ignoreAfterBar: logical, whether to split the tx id on the '|'\n",
       "          character to facilitate matching with the tx id in ‘tx2gene’\n",
       "          (default FALSE)\n",
       "\n",
       "geneIdCol: name of column with gene id. if missing, the ‘tx2gene’\n",
       "          argument can be used\n",
       "\n",
       " txIdCol: name of column with tx id\n",
       "\n",
       "abundanceCol: name of column with abundances (e.g. TPM or FPKM)\n",
       "\n",
       "countsCol: name of column with estimated counts\n",
       "\n",
       "lengthCol: name of column with feature length information\n",
       "\n",
       "importer: a function used to read in the files\n",
       "\n",
       "existenceOptional: logical, should tximport not check if files exist\n",
       "          before attempting import (default FALSE, meaning files must\n",
       "          exist according to ‘file.exists’)\n",
       "\n",
       "  sparse: logical, whether to try to import data sparsely (default is\n",
       "          FALSE). Initial implementation for ‘txOut=TRUE’,\n",
       "          ‘countsFromAbundance=\"no\"’ or ‘\"scaledTPM\"’, no inferential\n",
       "          replicates. Only counts matrix is returned (and abundance\n",
       "          matrix if using ‘\"scaledTPM\"’)\n",
       "\n",
       "sparseThreshold: the minimum threshold for including a count as a\n",
       "          non-zero count during sparse import (default is 1)\n",
       "\n",
       "readLength: numeric, the read length used to calculate counts from\n",
       "          StringTie's output of coverage. Default value (from\n",
       "          StringTie) is 75. The formula used to calculate counts is:\n",
       "          ‘cov * transcript length / read length’\n",
       "\n",
       "alevinArgs: named list, with logical elements ‘filterBarcodes’,\n",
       "          ‘tierImport’, ‘forceSlow’. See Details for definitions.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     *Inferential replicates:* ‘tximport’ will also load in information\n",
       "     about inferential replicates - a list of matrices of the Gibbs\n",
       "     samples from the posterior, or bootstrap replicates, per sample -\n",
       "     if these data are available in the expected locations relative to\n",
       "     the ‘files’. The inferential replicates, stored in ‘infReps’ in\n",
       "     the output list, are on estimated counts, and therefore follow\n",
       "     ‘counts’ in the output list. By setting ‘varReduce=TRUE’, the\n",
       "     inferential replicate matrices will be replaced by a single matrix\n",
       "     with the sample variance per transcript/gene and per sample.\n",
       "\n",
       "     *summarizeToGene:* While ‘tximport’ summarizes to the gene-level\n",
       "     by default, the user can also perform the import and summarization\n",
       "     steps manually, by specifing ‘txOut=TRUE’ and then using the\n",
       "     function ‘summarizeToGene’. Note however that this is equivalent\n",
       "     to ‘tximport’ with ‘txOut=FALSE’ (the default).\n",
       "\n",
       "     *Solutions on summarization:* regarding ‘\"tximport failed at\n",
       "     summarizing to the gene-level\"’:\n",
       "\n",
       "       1. provide a ‘tx2gene’ data.frame linking transcripts to genes\n",
       "          (more below)\n",
       "\n",
       "       2. avoid gene-level summarization by specifying ‘txOut=TRUE’\n",
       "\n",
       "     See ‘vignette('tximport')’ for example code for generating a\n",
       "     ‘tx2gene’ data.frame from a ‘TxDb’ object. The ‘tx2gene’\n",
       "     data.frame should exactly match and be derived from the same set\n",
       "     of transcripts used for quantifying (the set of transcript used to\n",
       "     create the transcriptome index).\n",
       "\n",
       "     *Tximeta:* One automated solution for Salmon or alevin\n",
       "     quantification data is to use the ‘tximeta’ function in the\n",
       "     tximeta Bioconductor package which builds upon and extends\n",
       "     ‘tximport’; this solution should work out-of-the-box for human and\n",
       "     mouse transcriptomes downloaded from GENCODE, Ensembl, or RefSeq.\n",
       "     For other cases, the user should create the ‘tx2gene’ manually as\n",
       "     shown in the tximport vignette.\n",
       "\n",
       "     *On tx2gene construction:* Note that the ‘keys’ and ‘select’\n",
       "     functions used to create the ‘tx2gene’ object are documented in\n",
       "     the man page for AnnotationDb-class objects in the AnnotationDbi\n",
       "     package (TxDb inherits from AnnotationDb). For further details on\n",
       "     generating TxDb objects from various inputs see\n",
       "     ‘vignette('GenomicFeatures')’ from the GenomicFeatures package.\n",
       "\n",
       "     *alevin:* The ‘alevinArgs’ argument includes some alevin-specific\n",
       "     arguments. This optional argument is a list with any or all of the\n",
       "     following named logical variables: ‘filterBarcodes’, ‘tierImport’,\n",
       "     and ‘forceSlow’. The variables are described as follows (with\n",
       "     default values in parens): ‘filterBarcodes’ (FALSE) import only\n",
       "     cell barcodes listed in ‘whitelist.txt’; ‘tierImport’ (FALSE)\n",
       "     import the tier information in addition to counts; ‘forceSlow’\n",
       "     (FALSE) force the use of the slower import R code even if\n",
       "     ‘fishpond’ is installed. For ‘type=\"alevin\"’ all arguments other\n",
       "     than ‘files’, ‘dropInfReps’, and ‘alevinArgs’ are ignored. Note\n",
       "     that ‘files’ should point to a single ‘quants_mat.gz’ file, in the\n",
       "     directory structure created by the alevin software (e.g. do not\n",
       "     move the file or delete the other important files). Note that\n",
       "     importing alevin quantifications will be much faster by first\n",
       "     installing the ‘fishpond’ package, which contains a C++ importer\n",
       "     for alevin's EDS format. For alevin, ‘tximport’ is importing the\n",
       "     gene-by-cell matrix of counts, as ‘txi$counts’, and effective\n",
       "     lengths are not estimated. ‘txi$mean’ and ‘txi$variance’ may also\n",
       "     be imported if inferential replicates were used, as well as\n",
       "     inferential replicates if these were output by alevin. Length\n",
       "     correction should not be applied to datasets where there is not an\n",
       "     expected correlation of counts and feature length.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     A simple list containing matrices: abundance, counts, length.\n",
       "     Another list element 'countsFromAbundance' carries through the\n",
       "     character argument used in the tximport call. The length matrix\n",
       "     contains the average transcript length for each gene which can be\n",
       "     used as an offset for gene-level analysis. If detected, and\n",
       "     ‘txOut=TRUE’, inferential replicates for each sample will be\n",
       "     imported and stored as a list of matrices, itself an element\n",
       "     ‘infReps’ in the returned list. An exception is alevin, in which\n",
       "     the ‘infReps’ are a list of bootstrap replicate matrices, where\n",
       "     each matrix has genes as rows and cells as columns. If\n",
       "     ‘varReduce=TRUE’ the inferential replicates will be summarized\n",
       "     according to the sample variance, and stored as a matrix\n",
       "     ‘variance’. alevin already computes the variance of the bootstrap\n",
       "     inferential replicates and so this is imported without needing to\n",
       "     specify ‘varReduce=TRUE’ (note that alevin uses the 1/N variance\n",
       "     estimator, so not the same as ‘var’).\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     Charlotte Soneson, Michael I. Love, Mark D. Robinson (2015)\n",
       "     Differential analyses for RNA-seq: transcript-level estimates\n",
       "     improve gene-level inferences. F1000Research. <URL:\n",
       "     http://doi.org/10.12688/f1000research.7563>\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     # load data for demonstrating tximport\n",
       "     # note that the vignette shows more examples\n",
       "     # including how to read in files quickly using the readr package\n",
       "     \n",
       "     library(tximportData)\n",
       "     dir <- system.file(\"extdata\", package=\"tximportData\")\n",
       "     samples <- read.table(file.path(dir,\"samples.txt\"), header=TRUE)\n",
       "     files <- file.path(dir,\"salmon\", samples$run, \"quant.sf.gz\")\n",
       "     names(files) <- paste0(\"sample\",1:6)\n",
       "     \n",
       "     # tx2gene links transcript IDs to gene IDs for summarization\n",
       "     tx2gene <- read.csv(file.path(dir, \"tx2gene.gencode.v27.csv\"))\n",
       "     \n",
       "     txi <- tximport(files, type=\"salmon\", tx2gene=tx2gene)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?tximport"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "10 \n",
      "11 \n",
      "12 \n",
      "13 \n",
      "14 \n",
      "15 \n",
      "16 \n",
      "17 \n",
      "18 \n",
      "19 \n",
      "20 \n",
      "21 \n",
      "22 \n",
      "23 \n",
      "24 \n",
      "\n",
      "\n",
      "transcripts missing from tx2gene: 1419\n",
      "\n",
      "summarizing abundance\n",
      "\n",
      "summarizing counts\n",
      "\n",
      "summarizing length\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Get tx2g\n",
    "tx2g <- data.table::fread(Sys.glob(glue::glue('/sc/hydra/projects/GuccioneLab/genome-indices/{genome}/ensembl/*-biomart.txt')))[,c(4, 1)]\n",
    "\n",
    "# Import\n",
    "txi <- tximport(infiles, type='kallisto', txOut=FALSE, tx2g=tx2g)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Note: importing `abundance.h5` is typically faster than `abundance.tsv`\n",
      "\n",
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "Warning message:\n",
      "“Duplicated column names deduplicated: '1147098' => '1147098_1' [3], '1147098' => '1147098_2' [4]”\n",
      "Warning message:\n",
      "“Unnamed `col_types` should have the same length as `col_names`. Using smaller of the two.”\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in tximport(infiles, type = \"kallisto\", txOut = TRUE): all(c(abundanceCol, countsCol, lengthCol) %in% names(raw)) is not TRUE\n",
     "output_type": "error",
     "traceback": [
      "Error in tximport(infiles, type = \"kallisto\", txOut = TRUE): all(c(abundanceCol, countsCol, lengthCol) %in% names(raw)) is not TRUE\nTraceback:\n",
      "1. tximport(infiles, type = \"kallisto\", txOut = TRUE)",
      "2. stopifnot(all(c(abundanceCol, countsCol, lengthCol) %in% names(raw)))"
     ]
    }
   ],
   "source": [
    "txi <- tximport(infiles, type='kallisto', txOut = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
